src/algorithms/pathfinders/dijkstra.ts

```diff
@@ -0,0 +1,111 @@
+import EdgeCost from '../models/edgecost';
+import Graph from '../models/graph';
+import Vertex from '../models/vertex';
+/**
+ * Implements Dijkstra's algorithm to find the shortest path between 2 vertices
+ */
+export default class DijkstraPathfinder {

```

Comment:
Have you thought about making the dijsktra algorithm part of the graph class itself? For me it would seem more logical as what you really do is performing some computation over the graph and making some output - exactly what a method does.

Reply:
I do not agree with this. There are some structures required (visited, ...) that should not be part of a Graph. Another reason is that if we make Dijkstra part of Graph then we could also make the skyline part of it and we end up with all the code inside one class.

I would keep it as is. The algorithm copies only vertex IDs which in my opinion is not that exhaustive.

Reply:
1. Mentioned structures should be inside the method only, not contained in the parent Graph class.
2. Inside the new method, other private methods like `producePath()` would be called.
3. This "class" is de facto a function, isn't it? We don't make classes that are used as functions. A class is something that acts as a container for then defined behaviour with specific properties, often described as something that can be compared to a real-life object (class Employee etc.). The key argument for this function is only the Graph class, one and only, so it should be logically inside this class (in theory, you could omit the vertices parameters and use first and last vertex of given graph - which wouldn't be much useful in real life, I know). To give an example, the method `Object.toString()` is a method of class Object, because it takes the object, perform computation over the object's property and outputs string. Pretty similar of what the dijkstra algorithm does (takes the graph, computes path and outputs an array). And we don't use `toString(object: Object)` either.
4. All code inside the class - yes, that could happen, but if it's logically correct, it should be fine. If you're concerned about the file length, some methods can be imported. Or if you really think that this function should be a separate class, it could be at least a child class of Graph.
5. Skyline part should not be part of the Graph class, as it has nothing to do with the Graph. It deals with already computed paths (arrays of vertices). Maybe we should make a class for these paths too.

Regarding the copying - I think you're right. I thought of it more deeply and you can't use the graph's vertices set as read-only, as you have to keep the distances and unvisited marker somewhere. So yes, it's not that exhaustive and it's actually needed.

I hope that you now have a clue why I think it shouldn't be in a separate class :D But I don't want to fight over it, so if you guys don't want to include it in the Graph, do it :)